be/src/exec/pipeline/operator.cpp

```diff
@@ -93,9 +97,57 @@ void Operator::eval_runtime_bloom_filters(vectorized::Chunk* chunk) {
     if (chunk == nullptr || chunk->is_empty()) {
         return;
     }
+
     if (auto* bloom_filters = runtime_bloom_filters()) {
         bloom_filters->evaluate(chunk, _bloom_filter_eval_context);
     }
+
+    eval_filter_null_values(chunk);
+}
+
+void Operator::eval_filter_null_values(vectorized::Chunk* chunk) {
+    const auto& filter_columns = filter_null_value_columns();
+    if (filter_columns.size() == 0) {
+        return;
+    }
+
+    size_t before_size = chunk->num_rows();
+    if (before_size == 0) {
+        return;
+    }
+
+    // lazy allocation.
+    vectorized::Buffer<uint8_t> selection(0);
+
+    for (SlotId slot_id : filter_columns) {
+        const ColumnPtr& c = chunk->get_column_by_slot_id(slot_id);
+        if (!c->is_nullable()) continue;
+        const vectorized::NullableColumn* nullable_column =
+                vectorized::ColumnHelper::as_raw_column<vectorized::NullableColumn>(c);
+        if (!nullable_column->has_null()) continue;
+        if (selection.size() == 0) {
+            selection.assign(before_size, 1);
+        }
+        // how many data() should we call? I really don't know.
+        // let compiler tells me.
+        // let compiler does vectorization.
+        // let compiler does everything.
+        // let's pray for compiler,
+        // till the end of the world.
+        const uint8_t* nulls = nullable_column->null_column()->raw_data();
+        uint8_t* sel = selection.data();
+        for (size_t i = 0; i < before_size; i++) {
+            sel[i] &= !nulls[i];
+        }
```

Comment:
Use vector[i] directly cannot auto-vectorization?

Reply:
Yes, `vector[i] ` can be auto-vectorization. It just a customary code.